[Bug Report] SVD tests fail on GPU

**Describe the bug**

`SVDInterpreter` tests fail on GPU.

The first issue is the tensors in the `assert`s are on different devices. This is easily fixed by sending all tensors to CPU before comparing.

Then we run into second issue:

SVD is not numerically stable so running on CPU gives different vectors than running on GPU. See https://pytorch.org/docs/stable/generated/torch.linalg.svd.html

The "correct" tensors in the test were probably computed on CPU so this fails on GPU.

**Proposed Solution**
Compare absolute values of singular vectors instead.

e.g.
```python
def test_svd_interpreter():
    svd_interpreter = SVDInterpreter(model)
    ov = svd_interpreter.get_singular_vectors(
        "OV", num_vectors=4, layer_index=0, head_index=0
    )
    w_in = svd_interpreter.get_singular_vectors(
        "w_in", num_vectors=4, layer_index=0, head_index=0
    )
    w_out = svd_interpreter.get_singular_vectors(
        "w_out", num_vectors=4, layer_index=0, head_index=0
    )

    ov = ov.abs()           #
    w_in = w_in.abs()       # ADD THESE LINES
    w_out = w_out.abs()     #

    ov, w_in, w_out = (
        ov.topk(2, dim=0).values,
        w_in.topk(2, dim=0).values,
        w_out.topk(2, dim=0).values,
    )
    assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
    assert torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
    assert torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
    assert torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
```

**Code example**

<details><summary>Details</summary>
<p>

```python
def test_svd_interpreter():
    svd_interpreter = SVDInterpreter(model)
    ov = svd_interpreter.get_singular_vectors(
        "OV", num_vectors=4, layer_index=0, head_index=0
    )
    w_in = svd_interpreter.get_singular_vectors(
        "w_in", num_vectors=4, layer_index=0, head_index=0
    )
    w_out = svd_interpreter.get_singular_vectors(
        "w_out", num_vectors=4, layer_index=0, head_index=0
    )
    ov, w_in, w_out = (
        ov.topk(2, dim=0).values,
        w_in.topk(2, dim=0).values,
        w_out.topk(2, dim=0).values,
    )
    assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
    assert torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
    assert torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
    assert torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
```

`make test`

```
>       assert torch.allclose(ov.to('cpu'), expected_OV_match, atol=ATOL)
E       AssertionError: assert False
E        +  where False = <built-in method allclose of type object at 0x7f61ebaf8420>(tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], grad_fn=<ToCopyBackward0>), tensor([[[0.6597, 0.8689, 0.5669, 0.7345]],\n\n        [[0.5232, 0.6705, 0.5623, 0.7240]]]), atol=0.0001)
E        +    where <built-in method allclose of type object at 0x7f61ebaf8420> = torch.allclose
E        +    and   tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], grad_fn=<ToCopyBackward0>) = <built-in method to of Tensor object at 0x7f620b63ad10>('cpu')
E        +      where <built-in method to of Tensor object at 0x7f620b63ad10> = tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], device='cuda:0',\n       grad_fn=<TopkBackward0>).to

```

</p>
</details> 


**System Info**
Installed from source on Ubuntu,
NVIDIA 4090, CUDA 10.1, Python 3.8

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/neelnanda-io/TransformerLens/issues) in the repo (**required**)

Thanks for pointing this out. Will accept fix if you write one or otherwise
solve soon.

On Tue, Jun 20, 2023, 10:21 PM Sviatoslav Chalnev ***@***.***>
wrote:

> *Describe the bug*
>
> SVDInterpreter tests fail on GPU.
>
> The first issue is the tensors in the asserts are on different devices.
> This is easily fixed by sending all tensors to CPU before comparing.
>
> Then we run into second issue:
>
> SVD is not numerically stable so running on CPU gives different vectors
> than running on GPU. See
> https://pytorch.org/docs/stable/generated/torch.linalg.svd.html
>
> The "correct" tensors in the test were probably computed on CPU so this
> fails on GPU.
>
> *Proposed Solution*
> Compare absolute values of singular vectors instead.
>
> e.g.
>
> def test_svd_interpreter():
>     svd_interpreter = SVDInterpreter(model)
>     ov = svd_interpreter.get_singular_vectors(
>         "OV", num_vectors=4, layer_index=0, head_index=0
>     )
>     w_in = svd_interpreter.get_singular_vectors(
>         "w_in", num_vectors=4, layer_index=0, head_index=0
>     )
>     w_out = svd_interpreter.get_singular_vectors(
>         "w_out", num_vectors=4, layer_index=0, head_index=0
>     )
>
>     ov = ov.abs()           #
>     w_in = w_in.abs()       # ADD THESE LINES
>     w_out = w_out.abs()     #
>
>     ov, w_in, w_out = (
>         ov.topk(2, dim=0).values,
>         w_in.topk(2, dim=0).values,
>         w_out.topk(2, dim=0).values,
>     )
>     assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
>     assert torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
>     assert torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
>     assert torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
>
> *Code example*
> Details
>
> def test_svd_interpreter():
>     svd_interpreter = SVDInterpreter(model)
>     ov = svd_interpreter.get_singular_vectors(
>         "OV", num_vectors=4, layer_index=0, head_index=0
>     )
>     w_in = svd_interpreter.get_singular_vectors(
>         "w_in", num_vectors=4, layer_index=0, head_index=0
>     )
>     w_out = svd_interpreter.get_singular_vectors(
>         "w_out", num_vectors=4, layer_index=0, head_index=0
>     )
>     ov, w_in, w_out = (
>         ov.topk(2, dim=0).values,
>         w_in.topk(2, dim=0).values,
>         w_out.topk(2, dim=0).values,
>     )
>     assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
>     assert torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
>     assert torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
>     assert torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
>
> make test
>
> >       assert torch.allclose(ov.to('cpu'), expected_OV_match, atol=ATOL)
> E       AssertionError: assert False
> E        +  where False = <built-in method allclose of type object at 0x7f61ebaf8420>(tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], grad_fn=<ToCopyBackward0>), tensor([[[0.6597, 0.8689, 0.5669, 0.7345]],\n\n        [[0.5232, 0.6705, 0.5623, 0.7240]]]), atol=0.0001)
> E        +    where <built-in method allclose of type object at 0x7f61ebaf8420> = torch.allclose
> E        +    and   tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], grad_fn=<ToCopyBackward0>) = <built-in method to of Tensor object at 0x7f620b63ad10>('cpu')
> E        +      where <built-in method to of Tensor object at 0x7f620b63ad10> = tensor([[[0.5244, 0.5373, 0.5669, 0.7077]],\n\n        [[0.5041, 0.5366, 0.5623, 0.6371]]], device='cuda:0',\n       grad_fn=<TopkBackward0>).to
>
>
> *System Info*
> Installed from source on Ubuntu,
> NVIDIA 4090, CUDA 10.1, Python 3.8
> Checklist
>
>    - I have checked that there is no similar issue
>    <https://github.com/neelnanda-io/TransformerLens/issues> in the repo (
>    *required*)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/neelnanda-io/TransformerLens/issues/328>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQPMYZ6ARV74ZA4EUZXNRSLXMGIU7ANCNFSM6AAAAAAZNGMNKQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/tests/unit/test_svd_interpreter.py
+++ b/tests/unit/test_svd_interpreter.py
@@ -49,9 +49,9 @@ def test_svd_interpreter():
         w_out.topk(2, dim=0).values,
     )
     assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
-    assert torch.allclose(ov, expected_OV_match, atol=ATOL)
-    assert torch.allclose(w_in, expected_w_in_match, atol=ATOL)
-    assert torch.allclose(w_out, expected_w_out_match, atol=ATOL)
+    assert torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
+    assert torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
+    assert torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
 
 
 def test_w_in_when_fold_ln_is_false():
@@ -60,7 +60,7 @@ def test_w_in_when_fold_ln_is_false():
         "w_in", num_vectors=4, layer_index=0, head_index=0
     )
     w_in = w_in.topk(2, dim=0).values
-    assert torch.allclose(w_in, expected_w_in_unfolded_match, atol=ATOL)
+    assert torch.allclose(w_in.cpu(), expected_w_in_unfolded_match, atol=ATOL)
 
 
 def test_svd_interpreter_returns_different_answers_for_different_layers():
@@ -81,9 +81,9 @@ def test_svd_interpreter_returns_different_answers_for_different_layers():
         w_out.topk(2, dim=0).values,
     )
     assert ov.shape == w_in.shape == w_out.shape == expected_OV_match.shape
-    assert not torch.allclose(ov, expected_OV_match, atol=ATOL)
-    assert not torch.allclose(w_in, expected_w_in_match, atol=ATOL)
-    assert not torch.allclose(w_out, expected_w_out_match, atol=ATOL)
+    assert not torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
+    assert not torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
+    assert not torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
 
 
 def test_svd_interpreter_returns_different_answers_for_different_models():
@@ -103,9 +103,9 @@ def test_svd_interpreter_returns_different_answers_for_different_models():
         w_in.topk(2, dim=0).values,
         w_out.topk(2, dim=0).values,
     )
-    assert not torch.allclose(ov, expected_OV_match, atol=ATOL)
-    assert not torch.allclose(w_in, expected_w_in_match, atol=ATOL)
-    assert not torch.allclose(w_out, expected_w_out_match, atol=ATOL)
+    assert not torch.allclose(ov.cpu(), expected_OV_match, atol=ATOL)
+    assert not torch.allclose(w_in.cpu(), expected_w_in_match, atol=ATOL)
+    assert not torch.allclose(w_out.cpu(), expected_w_out_match, atol=ATOL)
 
 
 # Failures

--- a/tests/unit/test_svd_interpreter.py
+++ b/tests/unit/test_svd_interpreter.py
@@ -14,19 +14,19 @@ second_model = HookedTransformer.from_pretrained("solu-3l")
 
 
 expected_OV_match = torch.Tensor(
-    [[[0.6597, 0.8689, 0.5669, 0.7345]], [[0.5232, 0.6705, 0.5623, 0.7240]]]
+    [[[0.6597, 0.8689, 0.6344, 0.7345]], [[0.5244, 0.6705, 0.5940, 0.7240]]]
 )
 
 expected_w_in_match = torch.Tensor(
-    [[[0.5572, 0.6466, 0.6406, 0.6094]], [[0.5417, 0.6103, 0.5773, 0.5726]]]
+    [[[0.7714, 0.6608, 0.6452, 0.6933]], [[0.7647, 0.6466, 0.6406, 0.6458]]]
 )
 
 expected_w_in_unfolded_match = torch.Tensor(
-    [[[0.2766, 0.3050, 0.3041, 0.3119]], [[0.2651, 0.2988, 0.2810, 0.2896]]]
+    [[[0.3639, 0.3164, 0.3095, 0.3430]], [[0.3614, 0.3050, 0.3041, 0.3140]]]
 )
 
 expected_w_out_match = torch.Tensor(
-    [[[0.5097, 0.4950, 0.5451, 0.7178]], [[0.5076, 0.4922, 0.5140, 0.7106]]]
+    [[[0.5097, 0.5389, 0.7906, 0.7178]], [[0.5076, 0.5350, 0.7674, 0.7106]]]
 )
 
 # Successes
@@ -36,13 +36,14 @@ def test_svd_interpreter():
     svd_interpreter = SVDInterpreter(model)
     ov = svd_interpreter.get_singular_vectors(
         "OV", num_vectors=4, layer_index=0, head_index=0
-    )
+    ).abs()
     w_in = svd_interpreter.get_singular_vectors(
         "w_in", num_vectors=4, layer_index=0, head_index=0
-    )
+    ).abs()
     w_out = svd_interpreter.get_singular_vectors(
         "w_out", num_vectors=4, layer_index=0, head_index=0
-    )
+    ).abs()
+
     ov, w_in, w_out = (
         ov.topk(2, dim=0).values,
         w_in.topk(2, dim=0).values,
@@ -58,7 +59,7 @@ def test_w_in_when_fold_ln_is_false():
     svd_interpreter = SVDInterpreter(unfolded_model)
     w_in = svd_interpreter.get_singular_vectors(
         "w_in", num_vectors=4, layer_index=0, head_index=0
-    )
+    ).abs()
     w_in = w_in.topk(2, dim=0).values
     assert torch.allclose(w_in.cpu(), expected_w_in_unfolded_match, atol=ATOL)
 
@@ -67,13 +68,13 @@ def test_svd_interpreter_returns_different_answers_for_different_layers():
     svd_interpreter = SVDInterpreter(model)
     ov = svd_interpreter.get_singular_vectors(
         "OV", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
     w_in = svd_interpreter.get_singular_vectors(
         "w_in", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
     w_out = svd_interpreter.get_singular_vectors(
         "w_out", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
 
     ov, w_in, w_out = (
         ov.topk(2, dim=0).values,
@@ -90,13 +91,13 @@ def test_svd_interpreter_returns_different_answers_for_different_models():
     svd_interpreter = SVDInterpreter(second_model)
     ov = svd_interpreter.get_singular_vectors(
         "OV", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
     w_in = svd_interpreter.get_singular_vectors(
         "w_in", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
     w_out = svd_interpreter.get_singular_vectors(
         "w_out", layer_index=1, num_vectors=4, head_index=0
-    )
+    ).abs()
 
     ov, w_in, w_out = (
         ov.topk(2, dim=0).values,

